packages/hardhat-core/src/internal/core/vars/vars-manager-setup.ts

```diff
@@ -0,0 +1,128 @@
+ *
+ * G = get function
+ * H = has function
+ * GD = get function with default value
+ *
+ * optional variables = H + (GD - G)
+ * required variables = G - H
+ */
+export class VarsManagerSetup extends VarsManager {
+  private readonly _getVarsAlreadySet: Set<string>;
+  private readonly _hasVarsAlreadySet: Set<string>;
+  private readonly _getVarsWithDefaultValueAlreadySet: Set<string>;
+
+  private readonly _getVarsToSet: Set<string>;
+  private readonly _hasVarsToSet: Set<string>;
+  private readonly _getVarsWithDefaultValueToSet: Set<string>;
+
+  constructor(varsFilePath: string) {
+    log("Creating a new instance of VarsManagerSetup");
+
+    super(varsFilePath);
+
+    this._getVarsAlreadySet = new Set();
+    this._hasVarsAlreadySet = new Set();
+    this._getVarsWithDefaultValueAlreadySet = new Set();
+
+    this._getVarsToSet = new Set();
+    this._hasVarsToSet = new Set();
+    this._getVarsWithDefaultValueToSet = new Set();
+  }
+
+  // Checks if the key exists, and updates sets accordingly.
+  // Ignore the parameter 'includeEnvs' defined in the parent class because during setup env vars are ignored.
+  public has(key: string): boolean {
+    log(`function 'has' called with key '${key}'`);
+
+    const hasKey = super.has(key);
+
+    if (hasKey) {
+      this._hasVarsAlreadySet.add(key);
+    } else {
+      this._hasVarsToSet.add(key);
+    }
+
+    return hasKey;
+  }
+
+  // Gets the value for the provided key, and updates sets accordingly.
+  // Ignore the parameter 'includeEnvs' defined in the parent class because during setup env vars are ignored.
+  public get(key: string, defaultValue?: string): string {
+    log(`function 'get' called with key '${key}'`);
+
+    const value = super.get(key, defaultValue);
+
+    if (value === undefined) {
+      this._getVarsToSet.add(key);
+      // Do not return undefined to avoid throwing an error
+      return "";
+    }
+
+    if (super.has(key) && defaultValue === undefined) {
+      this._getVarsAlreadySet.add(key);
+      return value;
+    }
+
+    // defaultValue is defined
+    if (super.has(key)) {
+      this._getVarsWithDefaultValueAlreadySet.add(key);
+    } else {
+      this._getVarsWithDefaultValueToSet.add(key);
+    }
+
+    return value;
+  }
+
+  public getRequiredVarsAlreadySet(): string[] {
+    return this._getRequired(this._getVarsAlreadySet, this._hasVarsAlreadySet);
+  }
+
+  public getOptionalVarsAlreadySet(): string[] {
+    return this._getOptionals(
+      this._getVarsAlreadySet,
+      this._hasVarsAlreadySet,
+      this._getVarsWithDefaultValueAlreadySet
+    );
+  }
+
+  public getRequiredVarsToSet(): string[] {
+    return this._getRequired(this._getVarsToSet, this._hasVarsToSet);
+  }
+
+  public getOptionalVarsToSet(): string[] {
+    return this._getOptionals(
+      this._getVarsToSet,
+      this._hasVarsToSet,
+      this._getVarsWithDefaultValueToSet
+    );
+  }
+
+  private _getRequired(getVars: Set<string>, hasVars: Set<string>): string[] {
```

Comment:
Let's add some comments explaining this method and the next one. I understand the reasoning now, but I'm pretty sure I'll forget in one month.